BIM: make ifcopenshell version upgrade more robust

furgo16 opened the pull request:
- Use Python `packaging`'s methods to compare semver version numbers. It's a dependency installed by default for the CAM workbench. This needs to be checked on different platforms/packages
- Allow upgrading packages by specifying the `--upgrade` option. Before, upgrade did not work if a version of IfcOpenShell was already installed on the system.
- Use `addonmanager_utilities`'s newly changed `create_pip_call` method. This allows the call to run on confined packages (snap) and the AppImage as well.

Other minor changes:
- Remove the now obsolete original method to compare versions
- The check for whether the pip version starts with "v" has also been removed. This no longer seems to be the case with recent releases: https://pypi.org/project/ifcopenshell/#history

## Known issues

- On Windows occasionally the "Unable to run pip. Please ensure pip is installed on your system" error appears. As pointed out at https://github.com/FreeCAD/FreeCAD/issues/19540 and in this PR's comments. This is not addressed in this PR, as it comes from Add-On Manager. It seems if there is a timeout calling `pip` for some reason, the process is killed and a signal is emitted. This PR does not handle any signals. See https://github.com/FreeCAD/FreeCAD/pull/19823#issuecomment-2685769487 and subsequent comments.

## Testing

1. As a shortcut to having to check out the branch and build FreeCAD, the changes are all confined in a single file that can be copied over to your FreeCAD installation => https://github.com/FreeCAD/FreeCAD/blob/031ac554226a9b89bc45097d517fa0610f7eaa83/src/Mod/BIM/nativeifc/ifc_openshell.py
2. Activate <kbd>Edit</kbd> > <kbd>Preferences...</kbd> > <kbd>General</kbd> > <kbd>Report view</kbd> > <kbd>Record log messages</kbd>
   ![Captura de pantalla de 2025-02-25 17-54-43](https://github.com/user-attachments/assets/c9154510-817b-4407-a9e3-2a19fd8c93e6)
4. Watch the Report panel for messages in case something goes wrong, and report any messages
5. Report your About info (<kbd>Help</kbd> > <kbd>About FreeCAD</kbd> > <kbd>Copy to clipboard</kbd>)

Fixes: https://github.com/FreeCAD/FreeCAD/issues/19821

furgo16 commented:
@semhustej would you have some time to test if this fixes the issue for you?

The changes are all confined in a single file that can be copied over => https://github.com/FreeCAD/FreeCAD/blob/031ac554226a9b89bc45097d517fa0610f7eaa83/src/Mod/BIM/nativeifc/ifc_openshell.py

furgo16 commented:
@semhustej thanks for testing. Could you try the instructions on the "Testing" section I've just added to the PR description? Essentially what you already did, but enabling recording of log messages.

furgo16 commented:
Thanks!

Invalid IfcOpenShell version detected: /.0.0 <= this is a known bug
Invalid IfcOpenShell version corrected to: 0.0.0 <= this is the workaround
Unable to run pip. Please ensure pip is installed on your system. <= I think here an error has happened and it's being masked by the "Unable to run pip" message.

Let me see if I can make that error more verbose.

furgo16 commented:
If you try again with the latest changes, the report view should give you the details of why the pip call failed.

furgo16 commented:
Sorry, I missed one bit thinking you were running into the wrong error. I now made all possible exceptions more verbose. Please try again with the latest change.

semhustej commented:
I am using this file, which was edited 30 minutes ago: https://github.com/FreeCAD/FreeCAD/blob/31c56ea1820626743964a6abb41266f7984472b7/src/Mod/BIM/nativeifc/ifc_openshell.py

The result is empty warning:
![empty-warning-02](https://github.com/user-attachments/assets/8a4702d6-b5db-4e71-bb0b-89a54a04f70c)

I also tried running in Safe mode with the same result.

furgo16 commented:
Apologies, try now again please.

semhustej commented:
No problem :)

Testing this file now:
![file-test](https://github.com/user-attachments/assets/6344cc02-3f42-49d5-b354-cc1ef22308c5)


The result is still no verbose output when updating IfcOpenShell
![outpit-03](https://github.com/user-attachments/assets/fa54c064-0fc1-499c-b865-aff2306323dc)

furgo16 commented:
Now I'm puzzled. I don't know why it's not even printing the error message. Perhaps the error comes from another masked exception somewhere else?

If you're still up for testing, could you try to run this macro to see if it produces a usable traceback? https://gist.github.com/furgo16/0105a307592ca4988eb29e4169680971 (remove the `.py` extension and put it in your macros folder, or create a new macro and copy over the contents).

Syres916 commented:
@furgo16 for info, a perfectly good macro that ran last week to update any Python dependency [macro](https://github.com/shaise/FreeCAD_SheetMetal/issues/461#issuecomment-2651160105) has stopped working already which says to me the many changes to the AddonManager are having unintended consequences, we need solid Unit Tests (first) to flush these out before merging otherwise people will become very frustrated being made to look like completely incompetent.

semhustej commented:
> Now I'm puzzled. I don't know why it's not even printing the error message. Perhaps the error comes from another masked exception somewhere else?
> 
> If you're still up for testing, could you try to run this macro to see if it produces a usable traceback? https://gist.github.com/furgo16/0105a307592ca4988eb29e4169680971 (remove the `.py` extension and put it in your macros folder, or create a new macro and copy over the contents).

I ran the macro and this is the output from the Report view window:

```

14:37:13  pyException: Traceback (most recent call last):
  File "C:/Users/semhu/AppData/Roaming/FreeCAD/Macro/ifc_openshell_lite.FCMacro", line 19, in <module>
    result = run_pip(args)
             ^^^^^^^^^^^^^
  File "C:/Users/semhu/AppData/Roaming/FreeCAD/Macro/ifc_openshell_lite.FCMacro", line 13, in run_pip
    result = utils.run_interruptable_subprocess(cmd)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\semhu\Documents\tyo\BF\FreeCAD\FC_daily\FreeCAD_weekly-builds-40317-conda-Windows-x86_64-py311\Mod\AddonManager\addonmanager_utilities.py", line 499, in run_interruptable_subprocess
    raise subprocess.CalledProcessError(
<class 'subprocess.CalledProcessError'>: Command '['C:\\Users\\semhu\\Documents\\tyo\\BF\\FreeCAD\\FC_daily\\FreeCAD_weekly-builds-40317-conda-Windows-x86_64-py311\\bin\\python.exe', '-m', 'pip', '--disable-pip-version-check', 'install', '--upgrade', '--disable-pip-version-check', '--target', 'C:\\Users\\semhu\\AppData\\Roaming\\FreeCAD\\Mod\\..\\AdditionalPythonPackages\\py311', 'ifcopenshell']' died with <Signals.CTRL_BREAK_EVENT: 1>.
```

furgo16 commented:
> <class 'subprocess.CalledProcessError'>: Command '['C:\\Users\\semhu\\Documents\\tyo\\BF\\FreeCAD\\FC_daily\\FreeCAD_weekly-builds-40317-conda-Windows-x86_64-py311\\bin\\python.exe', '-m', 'pip', '--disable-pip-version-check', 'install', '--upgrade', '--disable-pip-version-check', '--target', 'C:\\Users\\semhu\\AppData\\Roaming\\FreeCAD\\Mod\\..\\AdditionalPythonPackages\\py311', 'ifcopenshell']' died with <Signals.CTRL_BREAK_EVENT: 1>

Add-On Manager issue? The two references I've found after a quick search seem to point to it. https://forum.freecad.org/viewtopic.php?t=85164 and https://forum.freecad.org/viewtopic.php?t=75151, but I've ran out of time to look at it in detail today.

hyarion commented:
FYI, the ref-count bug has been merged now.

Syres916 commented:
> Out of interest, do the AppImages contain the pip binary? And do the native windows packages?

with reference to AppImages, here's the search of an extracted one, so I would say it's in usr/bin??:

![search_pip_appimage](https://github.com/user-attachments/assets/89b20cdb-01d0-45ab-81c0-43815141699b)


as far Windows (depends on how they are built I would guess) using the old LibPack kislore's builds have `pip.exe`, `pip3.exe` and `pip3.8.exe` included so yes in that case. The last Conda build I have is two years old and that had pip.exe and pip3.exe but can't comment on the latest builds.

Syres916 commented:
It won't take me long to determine what broke the macro and I'll report back later today.

semhustej commented:
> 
> @semhustej does running the Add-On Manager to install a random macro or workbench work for you?

I was able to install this macro from add-on manager without any reported errors:

`Macro 3DXML import was installed successfully`

Syres916 commented:
@furgo16 @semhustej the one area out of our control - the source appears to be the issue, I've just used my macro to install `networkx`, `pip` and `ezdxf` all at the first time of asking but `ifcopenshell` fails after some of it's dependencies like `numpy`, `pyparsing`, `fontTools` and `typing_extensions` are installed as expected


![PythonDependencies](https://github.com/user-attachments/assets/bbc58970-3c66-4d5b-a579-c8f5f9c8443a)

furgo16 commented:
@Syres916 is the failure the same as https://github.com/jmplonka/InventorLoader/issues/91#issuecomment-2558497328?

Unfortunately, I cannot reproduce it on snap. And the .deb daily packages have been broken for some time. I'll rebuild from source to test.

furgo16 commented:
I think at this point we'd be better off reporting an issue against Add-On Manager. I'd gladly do it, but I still cannot reproduce it. If someone else is willing to do it, happy to chime in as well.

Syres916 commented:
@furgo16 it must be a timeout, I cleared out my _py311_ folder and started going through installing each dependency, look what happened when I got to `scipy`, failed on first attempt and then when I added the line to output `result` it worked.

```
17:29:30  Install/Upgrade of pip finished
17:29:39  Install/Upgrade of lark finished
17:29:54  Install/Upgrade of fontTools finished
17:30:14  Install/Upgrade of typing-extensions finished
17:30:32  Install/Upgrade of numpy finished
17:30:47  Error trying to Install/Upgrade scipy. Please ensure pip is installed on your system and the module exists for this version of Python
17:31:19  Install/Upgrade of scipy finished
17:31:19  CompletedProcess(args=['/home/username/Downloads/FC_40176/squashfs-root/usr/bin/python', '-m', 'pip', 'install', '--disable-pip-version-check', '--target', '/home/username/.local/share/FreeCAD/Mod/../AdditionalPythonPackages/py311', 'scipy'], returncode=0, stdout='Collecting scipy\n  Using cached scipy-1.15.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\nCollecting numpy<2.5,>=1.23.5 (from scipy)\n  Using cached numpy-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)\nUsing cached scipy-1.15.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (37.6 MB)\nUsing cached numpy-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.4 MB)\nInstalling collected packages: numpy, scipy\nSuccessfully installed numpy-2.2.3 scipy-1.15.2\n', stderr="ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\nmatplotlib 3.10.0 requires fonttools>=4.22.0, which is not installed.\nmatplotlib 3.10.0 requires pyparsing>=2.3.1, which is not installed.\nWARNING: Target directory /home/username/.local/share/FreeCAD/AdditionalPythonPackages/py311/numpy-2.2.3.dist-info already exists. Specify --upgrade to force replacement.\nWARNING: Target directory /home/username/.local/share/FreeCAD/AdditionalPythonPackages/py311/numpy already exists. Specify --upgrade to force replacement.\nWARNING: Target directory /home/username/.local/share/FreeCAD/AdditionalPythonPackages/py311/numpy.libs already exists. Specify --upgrade to force replacement.\nWARNING: Target directory /home/username/.local/share/FreeCAD/AdditionalPythonPackages/py311/bin already exists. Specify --upgrade to force replacement.\n")
```

furgo16 commented:
Perhaps here via the ´p.kill()´? https://docs.python.org/3.10/library/subprocess.html#subprocess.Popen.kill

https://github.com/FreeCAD/FreeCAD/blob/56f0f640153926552effff9a3e931c8ce9a49e11/src/Mod/AddonManager/addonmanager_utilities.py#L470-L511

Syres916 commented:
@furgo16 I think I've cracked it, just need to retest a couple of times before I publish the steps, it's not going to be pretty!!

Syres916 commented:
@semhustej I'm working a little blind not having a Win10 box to hand but assuming the folder structure isn't a lot different from an AppImage as they are both Conda builds, the following steps should get somewhere:

With FreeCAD closed, navigate to `src/FreeCAD_weekly-builds-nnnnn-conda-Windows-x86_64-py311/bin/Lib/site-packages` and delete the `ifcopenshell` folder

navigate to `src/FreeCAD_weekly-builds-nnnnn-conda-Windows-x86_64-py311/Mod/BIM` and edit `ifc_openshell.py`, change line 107 from:

`        args = ["install", "--disable-pip-version-check", "--target", vendor_path, "ifcopenshell"]`

to:

`        args = ["install", "--disable-pip-version-check", "--target", vendor_path, "ifcopenshell==0.8.0", "--upgrade"]`

and save the file.

Leave the following folder open to check changes `C:/Users/username/AppData/Roaming/FreeCAD/AdditionalPythonPackages/py311`

Start FreeCAD and switch to BIM Wb click `Utils` menu and click `IfcOpenShell update` and _**hopefully**_ you don't get an error now and you'll see new folders in your py311 folder!!

semhustej commented:
@Syres916 
I will test this tomorrow. 

Should I edit the default `ifc_openshell.py` or use the `ifc_openshell.py` provided by furgo16 in this pull request?

Syres916 commented:
> Should I edit the default `ifc_openshell.py` or use the `ifc_openshell.py` provided by furgo16 in this pull request?

Sorry, you can use the one in this PR but just change the end of line 109 args from `"ifcopenshell"` to `"ifcopenshell==0.8.0"` Basically there's something with Python 3.11 that's not behaving with the later `ifcopenshell` versions so we're forcing it back to 0.8.0, Python 3.12 seems to be fine with the latest.

furgo16 commented:
> FYI, the ref-count bug has been merged now.

@hyarion I saw that thanks! While the fix is now available on the snap daily builds, I'll wait until the it is also included in the AppImage weekly builds before I remove the workaround snippet on this PR.

furgo16 commented:
> EDIT: the update information window still tells me it will install 0.8.0post1 version of IOS.

That's to be expected.

The original code:
1. First looks for the highest version of ifcopenshell online and reports it to the user
2. Then tries to install (not upgrade) ifcopenshell

The code in this PR:
1. First looks for the highest version of ifcopenshell online and reports it to the user (same as the original)
2. Then tries to install (and upgrade if necessary) ifcopenshell

The modified code from `@Syres` which you've just tested:
1. First looks for the highest version of ifcopenshell online and reports it to the user (same as the original)
2. Then tries to install (and upgrade if necessary) ifcopenshell to 0.8.0, regardless of what was found on step 1

furgo16 commented:
@Syres916 thanks for the investigation!

> Basically there's something with Python 3.11 that's not behaving with the later ifcopenshell versions so we're forcing it back to 0.8.0, Python 3.12 seems to be fine with the latest.

If I understand it correctly, there is some issue with `Python 3.11.9` (and pip, I guess?) when downloading ifcopenshell `0.8.1post1`. That issue is then propagated to Add-On Manager and causes the timeout. Which in turn kills the `pip` process and emits the signal, which we don't handle on this PR.

If that's the case, I feel this should probably be addressed in Add-On Manager itself. My suggestion would be to leave this PR as it is, while addressing any other feedback for improvements. Signal interception should probably be done higher up the stack.

semhustej commented:
UPDATE: i previously did not do all the requested steps. Now I did according to @Syres916 instructions. This is the result:

- the IOS update shows empty error in the Report view
- after the update, version 0.8.0 is reported:
![info-after-update](https://github.com/user-attachments/assets/2bbb7f66-acc9-433a-9df0-018da932b8d7)




I did not see any change in the py311 folder though:

**Before:**
![before-folder](https://github.com/user-attachments/assets/af63901d-cc6f-44be-885e-6f21d2947243)



**After:**
![after-folder](https://github.com/user-attachments/assets/380c391b-3314-44f5-9241-0ce306ec25f9)

yorikvanhavre commented:
In any case, I think we would do good to always report any version problem to ifcopenshell. That would help them to get conscious about it and that we rely on it ;)

It's one of the things that can help them shape things better for wider distribution and inclusion (ex. in linux distributions repos). That would do good to us too.

furgo16 commented:
IfcOpenShell version corruption workaround removed, now that https://github.com/FreeCAD/FreeCAD/pull/19861 has landed on the weekly builds as well. Updated PR description accordingly and added "Known issues" section.

furgo16 commented:
> In any case, I think we would do good to always report any version problem to ifcopenshell. That would help them to get conscious about it and that we rely on it ;)

I fully agree, that's why I pinged upstream here: https://github.com/FreeCAD/FreeCAD/issues/19821#issuecomment-2678514426. Although it ended up being an issue of ours :)

As per reporting the Python 3.11 vs ifcopenshell 0.8.1post1 issue, I'll let `@Syres916 `, who did the debugging and has the knowledge, decide.

Syres916 commented:
@yorikvanhavre I've logged an [issue](https://github.com/IfcOpenShell/IfcOpenShell/issues/6253) with ifcopenshell but the lack of error message isn't going to help much.

yorikvanhavre commented:
@furgo16 is this good for merge now? It seems to me it is, but just wanted to make sure :)
